### PR TITLE
Add AddCommandParserTest #327

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Part of #327

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/3] [Rename EditCommandTestUtil to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/80a8304d593c11b6370f0fafe19e8ab89830f0db)
* [v1 2/3] [Extract constants from EditCommandParserTest to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/0ccfce8ee5e10147e98efd6c854a58372cc6f484)
* [v1 3/3] [Add tests for AddCommandParser](https://github.com/se-edu/addressbook-level4/pull/494/commits/15e75d78dbb1627456b57275bdb68f64d7c457fa)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...15e75d78dbb1627456b57275bdb68f64d7c457fa))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/494/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        assertParseFailure(AddCommand.COMMAND_WORD + " " + NAME_DESC_BOB + PHONE_DESC_BOB
+                        + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + INVALID_TAG_DESC + VALID_TAG_FRIEND,
+                Tag.MESSAGE_TAG_CONSTRAINTS);
+
+        // two invalid values
+        assertParseFailure(AddCommand.COMMAND_WORD + " " + INVALID_NAME_DESC + PHONE_DESC_BOB
+                        + EMAIL_DESC_BOB + INVALID_ADDRESS_DESC, Name.MESSAGE_NAME_CONSTRAINTS);
+    }
+
+    private void assertParseSuccess(String userInput, Person expectedPerson) throws Exception {
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Needs a header comment for this method.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                .withTags(VALID_TAG_FRIEND).build();
+        assertParseSuccess(AddCommand.COMMAND_WORD + " " + NAME_DESC_BOB + PHONE_DESC_AMY + PHONE_DESC_BOB
+                + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_FRIEND, person);
+    }
+
+    @Test
+    public void parse_invalidArgsFormat_failure() {
+        String expectedMessage = String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE);
+
+        // missing parameters
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: // missing **mandatory** parameters

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_HUSBAND, person);
+    }
+
+    @Test
+    public void parse_multipleSequentialTags_success() throws Exception {
+        Person person = new PersonBuilder().withName(VALID_NAME_BOB).withPhone(VALID_PHONE_BOB)
+                .withEmail(VALID_EMAIL_BOB).withAddress(VALID_ADDRESS_BOB)
+                .withTags(VALID_TAG_FRIEND, VALID_TAG_HUSBAND).build();
+        assertParseSuccess(AddCommand.COMMAND_WORD + " " + NAME_DESC_BOB + PHONE_DESC_BOB
+                + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_HUSBAND + TAG_DESC_FRIEND, person);
```
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: What about duplicate tags? E.g. `+ TAG_DESC_FRIEND + TAG_DESC_FRIEND`

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    @Test
+    public void parse_twoPhones_success() throws Exception {
```
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Maybe `parse_twoPhones_lastPhoneAccepted()` would be better?

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    public void setUp() {
+        model = new ModelManager();
+    }
+
+    @Test
+    public void parse_spacesInInput_success() throws Exception {
+        Person person = new PersonBuilder().withName(VALID_NAME_AMY).withPhone(VALID_PHONE_AMY)
+                .withEmail(VALID_EMAIL_AMY).withAddress(VALID_ADDRESS_AMY).withTags().build();
+        assertParseSuccess("   " + AddCommand.COMMAND_WORD + "  " + NAME_DESC_AMY + "   "
+                + PHONE_DESC_AMY + "   " + EMAIL_DESC_AMY + "   " + ADDRESS_DESC_AMY + "    ", person);
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Do you need to put spaces between two fields? (e.g. `NAME_DESC_AMY` and `PHONE_DESC_AMY`?) They already have a space attached at the front:

```java
public static final String NAME_DESC_BOB = " " + PREFIX_NAME + VALID_NAME_BOB;
```

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    @Test
+    public void parse_noTags_success() throws Exception {
+        Person person = new PersonBuilder().withName(VALID_NAME_AMY).withPhone(VALID_PHONE_AMY)
+                .withEmail(VALID_EMAIL_AMY).withAddress(VALID_ADDRESS_AMY).withTags().build();
+        assertParseSuccess(AddCommand.COMMAND_WORD + " " + NAME_DESC_AMY + PHONE_DESC_AMY
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Same here, the space doesn't seem necessary.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        assertParseFailure(AddCommand.COMMAND_WORD + " " + VALID_NAME_AMY + PHONE_DESC_BOB
+                + EMAIL_DESC_BOB + ADDRESS_DESC_BOB, expectedMessage);
+    }
+
+    @Test
+    public void parse_invalidValue_failure() {
+        // invalid name
+        assertParseFailure(AddCommand.COMMAND_WORD + " " + INVALID_NAME_DESC + PHONE_DESC_BOB
+                        + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_HUSBAND + TAG_DESC_FRIEND,
+                Name.MESSAGE_NAME_CONSTRAINTS);
```
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Indentation here looks weird. Maybe line up this line with the previous? Same for other fields.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        assertParseFailure(AddCommand.COMMAND_WORD + " " + NAME_DESC_BOB + PHONE_DESC_BOB
+                        + EMAIL_DESC_BOB + INVALID_ADDRESS_DESC + TAG_DESC_HUSBAND + TAG_DESC_FRIEND,
+                Address.MESSAGE_ADDRESS_CONSTRAINTS);
+
+        // invalid tag
+        assertParseFailure(AddCommand.COMMAND_WORD + " " + NAME_DESC_BOB + PHONE_DESC_BOB
+                        + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + INVALID_TAG_DESC + VALID_TAG_FRIEND,
+                Tag.MESSAGE_TAG_CONSTRAINTS);
+
+        // two invalid values
```
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Mention that only one invalid value is reported.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 1 above)
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Header comments for assertParseSuccess(...) and assertParseFailure(...) missing.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        command.setData(model, new CommandHistory());
+        command.execute();
+
+        assertEquals(expectedPerson, model.getAddressBook().getPersonList().get(0));
+    }
+
+    private void assertParseFailure(String userInput, String expectedMessage) {
+        try {
+            parser.parse(userInput);
+            fail();
```
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: The expected ParseException was not thrown.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 1 above)
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Same for assertParseFailure

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/3] [Rename EditCommandTestUtil to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/32f3f5767b5a1b420e48a81982587037afbedf13)
* [v2 2/3] [Extract constants from EditCommandParserTest to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/875e47e67ecc49f42a53e5fa1b80f2284d37c16a)
* [v2 3/3] [Add tests for AddCommandParser](https://github.com/se-edu/addressbook-level4/pull/494/commits/118588b0790c27007a14e8c887143e1e8c3a8cf4)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...118588b0790c27007a14e8c887143e1e8c3a8cf4)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/7029d6439920ae506fbeafbd7b2c83521061aba4~1...7029d6439920ae506fbeafbd7b2c83521061aba4))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/494/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@yamgent @limmlingg review pls!

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    private AddCommandParser parser = new AddCommandParser();
+    private Model model;
+
+    @Before
+    public void setUp() {
+        model = new ModelManager();
+    }
+
+    @Test
+    public void parse_spacesInInput_success() throws Exception {
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: `parse_multipleSpacesInInput_success()`

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 7 above)
Comment (yamgent) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Indentation issue not resolved. We should not use indentation purely to indicate that they are two different parameters.

Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: This is IntelliJ's auto indentation.

Comment (yamgent) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: > This is IntelliJ's auto indentation.

Ok, @limmlingg do take note.

Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Note taken.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                        + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + INVALID_TAG_DESC + VALID_TAG_FRIEND,
+                Tag.MESSAGE_TAG_CONSTRAINTS);
+
+        // two invalid values, only first invalid value reported
+        assertParseFailure(AddCommand.COMMAND_WORD + INVALID_NAME_DESC + PHONE_DESC_BOB
+                        + EMAIL_DESC_BOB + INVALID_ADDRESS_DESC, Name.MESSAGE_NAME_CONSTRAINTS);
+    }
+
+    /**
+     * Asserts the parsing of {@code userInput} is successful and the person created
```
Comment (limmlingg) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Asserts **that** the parsing...

Same for assertParseFailure

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/3] [Rename EditCommandTestUtil to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/32f3f5767b5a1b420e48a81982587037afbedf13)
* [v3 2/3] [Extract constants from EditCommandParserTest to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/875e47e67ecc49f42a53e5fa1b80f2284d37c16a)
* [v3 3/3] [Add tests for AddCommandParser](https://github.com/se-edu/addressbook-level4/pull/494/commits/689779f887ec404667d075ce98994b74ffd0e383)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...689779f887ec404667d075ce98994b74ffd0e383)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/09e28c8d7e2b1773a10e5154135c456c4e8c194b~1...09e28c8d7e2b1773a10e5154135c456c4e8c194b))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/494/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@limmlingg

File: src/test/java/seedu/address/testutil/CommandTestUtil.java [hunk 12]
```diff
@@ -18,6 +24,22 @@
...
+    public static final String PHONE_DESC_AMY = " " + PREFIX_PHONE + VALID_PHONE_AMY;
+    public static final String PHONE_DESC_BOB = " " + PREFIX_PHONE + VALID_PHONE_BOB;
+    public static final String EMAIL_DESC_AMY = " " + PREFIX_EMAIL + VALID_EMAIL_AMY;
+    public static final String EMAIL_DESC_BOB = " " + PREFIX_EMAIL + VALID_EMAIL_BOB;
+    public static final String ADDRESS_DESC_AMY = " " + PREFIX_ADDRESS + VALID_ADDRESS_AMY;
+    public static final String ADDRESS_DESC_BOB = " " + PREFIX_ADDRESS + VALID_ADDRESS_BOB;
+    public static final String TAG_DESC_FRIEND = " " + PREFIX_TAG + VALID_TAG_FRIEND;
+    public static final String TAG_DESC_HUSBAND = " " + PREFIX_TAG + VALID_TAG_HUSBAND;
+
+    public static final String INVALID_NAME_DESC = " " + PREFIX_NAME + "James&"; // '&' not allowed in names
```
Comment (chao1995) [inline src/test/java/seedu/address/testutil/CommandTestUtil.java]: The naming convention of `NAME_DESC_AMY` and `INVALID_NAME_DESC` doesn't seem to match. Maybe `INVALID_NAME_DESC` should be `NAME_DESC_INVALID`? Perhaps can be fixed in a different commit though.

Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/testutil/CommandTestUtil.java]: I think this should be fixed inside another PR, as refactoring this out would cause a lot of diff :<

Comment (yamgent) [inline src/test/java/seedu/address/testutil/CommandTestUtil.java]: Should create a new issue for it.

**Comment by Zhiyuan-Amos** [comment]
@chao1995 ah what you said makes sense. It seems like I did too much black-box testing, will make the changes :>

**Comment by Zhiyuan-Amos** [comment]
Actually I noticed that I did this PR wrongly :<

I should have implemented `#equals(Object)` method for `AddCommand`. That way, I don't have to test using the `Model`.

**Comment by CanIHasReview** [comment]
# v4

@Zhiyuan-Amos submitted v4 for review.

* [v4 1/4] [Rename EditCommandTestUtil to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/40fa3753c1ec520b0ff8fd8eab3b5b22d074ca32)
* [v4 2/4] [Extract constants from EditCommandParserTest to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/fb1d73bde3e5b20d393d3bf9d48e58f35c77cbd3)
* [v4 3/4] [AddCommand: Implement equals\(Object\) method](https://github.com/se-edu/addressbook-level4/pull/494/commits/2d1c6e91359ef44eab66d07507f80ae412e39770)
* [v4 4/4] [Add tests for AddCommandParser](https://github.com/se-edu/addressbook-level4/pull/494/commits/359f9915feccf10b72b17bd46a08e772eb11d18b)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/97653b7fa08a90902a9a5c8b800007d3bbb483bf...359f9915feccf10b72b17bd46a08e772eb11d18b)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/7999f8a9042142e593feecbe57d66a7345383528~1...7999f8a9042142e593feecbe57d66a7345383528))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/494/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+    @Test
+    public void parse_multipleTags_success() throws Exception {
+        Person expectedPerson = new PersonBuilder().withName(VALID_NAME_BOB).withPhone(VALID_PHONE_BOB)
+                .withEmail(VALID_EMAIL_BOB).withAddress(VALID_ADDRESS_BOB)
+                .withTags(VALID_TAG_FRIEND, VALID_TAG_HUSBAND).build();
+        assertParseSuccess(AddCommand.COMMAND_WORD + NAME_DESC_BOB + PHONE_DESC_BOB
+                + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_HUSBAND + TAG_DESC_FRIEND,
+                new AddCommand(expectedPerson));
+    }
```
Comment (damithc) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: To have similar granularity in how we divide code into individual tests, may be something like this?
* parse_allFieldsPresent_success (should include multiple tags)
* parse_optionalFieldsMissing_success (covers case: 0 tags)
* parse_compulsoryFieldMissing_failure
* parse_invalidValue_failure

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+import seedu.address.model.person.Address;
+import seedu.address.model.person.Email;
+import seedu.address.model.person.Name;
+import seedu.address.model.person.Person;
+import seedu.address.model.person.Phone;
+import seedu.address.model.tag.Tag;
+import seedu.address.testutil.PersonBuilder;
+
+/**
+ * Unit and Integration tests for {@code AddCommandParser}. {@code AddCommandParser}'s unit tests
```
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Yes, I will be removing the header comments. Initially, I wanted to verify the correctness of `AddCommandParser#parse(String)` by using the Model. Thus, I decided to write header comments to explain why so (as the other parsers are tested using only unit tests, not hybrid tests).

However, I noticed that I can actually use `AddCommand#equals(Object)` to verify the correctness.

As such, I will remove the header comments.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        Person expectedPerson = new PersonBuilder().withName(VALID_NAME_BOB).withPhone(VALID_PHONE_BOB)
+                .withEmail(VALID_EMAIL_BOB).withAddress(VALID_ADDRESS_BOB)
+                .withTags(VALID_TAG_FRIEND, VALID_TAG_HUSBAND).build();
+        assertParseSuccess(AddCommand.COMMAND_WORD + NAME_DESC_BOB + PHONE_DESC_BOB
+                + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_HUSBAND + TAG_DESC_FRIEND,
+                new AddCommand(expectedPerson));
+    }
+
+    @Test
+    public void parse_multipleArgs_lastArgAccepted() throws Exception {
```
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: I think this should be handled by the SUT.
<pre>
Name name = ParserUtil.parseName(argMultimap.<b>getValue</b>(PREFIX_NAME)).get();
</pre>

If the function called is `argMultimap.getFirstValue(...)`, the result would be different.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 13 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Yup, looks tidier this way.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        assertParseFailure(AddCommand.COMMAND_WORD + NAME_DESC_BOB + VALID_PHONE_BOB
+                + EMAIL_DESC_BOB + ADDRESS_DESC_BOB, expectedMessage);
+
+        // missing email prefix
+        assertParseFailure(AddCommand.COMMAND_WORD + NAME_DESC_BOB + PHONE_DESC_BOB
+                + VALID_EMAIL_BOB + ADDRESS_DESC_BOB, expectedMessage);
+
+        // missing address prefix
+        assertParseFailure(AddCommand.COMMAND_WORD + NAME_DESC_BOB + PHONE_DESC_BOB
+                + EMAIL_DESC_BOB + VALID_ADDRESS_BOB, expectedMessage);
```
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Noted. I suppose this is to test the private method `arePrefixesPresent(...)`?

**Comment by CanIHasReview** [comment]
# v5

@Zhiyuan-Amos submitted v5 for review.

* [v5 1/4] [Rename EditCommandTestUtil to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/a4bc9f8b32900dc0dfd43acf573c586c329e4b59)
* [v5 2/4] [Extract constants from EditCommandParserTest to CommandTestUtil](https://github.com/se-edu/addressbook-level4/pull/494/commits/cc2b06cdcf78ec88b425fadd4470dc4e5a065627)
* [v5 3/4] [AddCommand: Implement equals\(Object\) method](https://github.com/se-edu/addressbook-level4/pull/494/commits/c4e0a604808d195dcec7b7e933e6b0d275eec2e4)
* [v5 4/4] [Add tests for AddCommandParser](https://github.com/se-edu/addressbook-level4/pull/494/commits/b1ab21f8350f16616af46922255e7794f2798d0e)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/d4dd4c253b3218faf475d5d3941e5db377e62637...b1ab21f8350f16616af46922255e7794f2798d0e)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/c997192a4b86a561ccd447a6e342fda087a70604~1...c997192a4b86a561ccd447a6e342fda087a70604))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/494/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>
